content/helpers.go

```diff
@@ -90,38 +120,65 @@ func WriteBlob(ctx context.Context, cs Ingester, ref string, r io.Reader, desc o
 // OpenWriter opens a new writer for the given reference, retrying if the writer
 // is locked until the reference is available or returns an error.
 func OpenWriter(ctx context.Context, cs Ingester, opts ...WriterOpt) (Writer, error) {
-	var (
-		cw    Writer
-		err   error
-		retry = 16
-	)
-	for {
-		cw, err = cs.Writer(ctx, opts...)
-		if err != nil {
-			if !errdefs.IsUnavailable(err) {
-				return nil, err
-			}
+	wOpt := &WriterOpts{}
+	for _, o := range opts {
+		o(wOpt)
+	}
+
+	ref := wOpt.Ref
+	if ref == "" {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "ref must not be empty")
+	}
 
-			// TODO: Check status to determine if the writer is active,
-			// continue waiting while active, otherwise return lock
-			// error or abort. Requires asserting for an ingest manager
+	// TODO: Keep the ref with namespace scope.
+	if err := keyLocks.lock(ctx, ref); err != nil {
+		return nil, err
+	}
 
-			select {
-			case <-time.After(time.Millisecond * time.Duration(rand.Intn(retry))):
-				if retry < 2048 {
-					retry = retry << 1
+	// Even if the lock has been hold, the lock is on one client side and it
+	// still receives unavailable error when other client already holds the
+	// reference in the content backend. We still needs backoff retry here.
+	cw, err := func() (Writer, error) {
+		var (
+			cw    Writer
+			err   error
+			retry = 16
+		)
+
+		for {
+			cw, err = cs.Writer(ctx, opts...)
```

Comment:
shall we provide a new method like cs.WriterWithLock here as the lock by ref logic has been moved to OpenWriter, so there should be no lock contention inside Writer method